[MRG, API] Start of automatic methods section with create_methods_paragraph

adam2392 opened the pull request:
PR Description
--------------

Addresses preliminarily: #347 

**A summary of MEEG summary (MEG, EEG, iEEG):**
I don't want this to be too much headache to start with, so figured that the easiest most robust summary we can provide is modularized as such:

1. **dataset description:** subject, session, kinds, and dataset_description.json file
2. **`participants.tsv` file summary per subject:** age, sex, hand that is supported in mne-bids. Note this file is only **RECOMMENDED**. I went through a lot of effort to get this to work without making the report look crazy ugly for now because this I figured is one of the most crucial summaries every study should have, but since it's structure is not very imposed by BIDS, then it's hard to summarize consistently.
3. **modality-agnostic-summary:** per session scans, and their length, sfreq, channel counts, etc.
4. ~~**modality-specific-summary:** adding iEEG channel counts (e.g. SEEG, ECoG, etc.). Similarly, I suppose if someone wants to add MEG/EEG, it can be a relatively short summary here.~~ (tabled to future)

**TODO:**

- [x] Update `convert_group_studies` to use `create_methods_paragraph`.
- [x] Have maintainers first review the groundwork to make sure this is in the right direction.
- [x] Add documentation to the docstrings
- [x] Add a summary function for channels.tsv and sidecar.json files ~~`kind=ieeg` data (idk how to add for 'eeg', or 'meg', so would prefer someone else add that functionality in)~~
- [x] Add REQUIRED elements from `dataset_description.json` and the reference/DOI for mne-bids.
- [x] Add example outputs from OpenNeuro datasets (i.e. iEEG)
- [ ] How to deal w/ `emptyroom` subjects? I don't work w/ these, so skipping them for now. Assuming this is okay, I added a `XXX` to the inline comments for someone else to fix.
- [ ] How to add MEG/iEEG/EEG specific data, such as "Gradiometers" and "Magnometers" for MEG data?
- [ ] Adding scan-level summary even without the `*_scans.tsv` files, which is considered "RECOMMENDED" and not "REQUIRED".

**Example Output from Local/OpenNeuro datasets**
These are the datasets I ran the method generation w/:

0. local_dataset I have, that will get put onto openneuro.
1. ds001779,
2. ds002778
3. ds002904
4. ds000246': 'https://github.com/OpenNeuroDatasets/ds000246',
5. ds000248': 'https://github.com/OpenNeuroDatasets/ds000248',
6. ds000117': 'https://github.com/OpenNeuroDatasets/ds000117',
7. ds001810': 'https://github.com/OpenNeuroDatasets/ds001810',
8. ds001971': 'https://github.com/OpenNeuroDatasets/ds001971',
9. somato

Ran datasets w/ the following code:

```Python
methods_paragraph = create_methods_paragraph(bids_root)
print(methods_paragraph)
```

See output on PR here: `./report.txt`.

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"
- [ ] Commit history does not contain any merge commits

jasmainak commented:
nice, can you share an example paragraph generated by this report?

jasmainak commented:
maybe update the ds000117 example?

sappelhoff commented on mne_bids/tests/test_report.py:
```diff
@@ -0,0 +1,60 @@
+
+import pytest
+
+with warnings.catch_warnings():
+    warnings.filterwarnings(action='ignore',
+                            message="can't resolve package",
+                            category=ImportWarning)
+    import mne
```
what's this about?

adam2392 replied:
copied and pasted from another file. :p

adam2392 commented:
> maybe update the ds000117 example?

Done! lmk what you think.

> nice, can you share an example paragraph generated by this report?

It is copied into the PR description.

jasmainak commented:
> There are 2 datasets (364.69 +/- 116.31 seconds) with sampling rates 1000.0 (n=1), 999.0 (n=1).

I don't get this part.

Can you make the text almost like a copy-paste? I would also add information about:

1. Manufacturer of device (e.g., Vectorview)
2. Number of channels
3. Filtering information

what's the output of `pybids` for the same dataset? hopefully not the same?

jasmainak commented:
This is the output for ds000117:

> The dataset consists of 1 patients with 1 sessions (01) consisting of 1 kinds of data (meg). The dataset consists of 1 subjects (10.0 +/- 0.0; 1 right ; 1 male ). There are 6 datasets (447.33 +/- 106.20 seconds) with sampling rates 1100.0 (n=6)

This seems off. There aren't "6 datasets". We should test on a few datasets to ensure it gives something reasonable. I would pick 10 ephys datasets at random from openneuro and try on them.

adam2392 commented:
> > There are 2 datasets (364.69 +/- 116.31 seconds) with sampling rates 1000.0 (n=1), 999.0 (n=1).
> Can you make the text almost like a copy-paste for publication?

What would be the copy-paste version of this part? Do you have a specific structure in mind?

> 
> 1. Manufacturer of device (e.g., Vectorview)
> 2. Number of channels
> 3. Filtering information
> 4. Sampling frequency

By filtering info, I suppose you mean the `SoftwareFilters` in sidecar.json? 1/2/4 can be added. 

> what's the output of `pybids` for the same dataset? hopefully not the same?

`pybids` doesn't give any output at all. See the output pasted in the corresponding issue of this PR. I think it's due to the fact that the pybids-report still only supports nifti files.

sappelhoff commented on examples/convert_group_studies.py:
```diff
@@ -100,3 +100,8 @@
+###############################################################################
+# Now let's generate a report on the dataset.
+dataset_summary = create_methods_paragraph(bids_root=output_path)
+print(dataset_summary)
```
nitpick: consider using [textwrap](https://docs.python.org/3/library/textwrap.html) for pretty printing this

![image](https://user-images.githubusercontent.com/9084751/85435440-bdef2400-b587-11ea-8bf8-918bb654a7ba.png)

adam2392 commented:
> This is the output for ds000117:
> This seems off. There aren't "6 datasets". We should test on a few datasets to ensure it gives something reasonable. I would pick 10 ephys datasets at random from openneuro and try on them.

https://mne.tools/mne-bids/stable/auto_examples/convert_group_studies.html#sphx-glr-auto-examples-convert-group-studies-py

There are 6 `.fif` files tho, so aren't there 6 datasets?

agramfort commented:
@adam2392 please paste here what you obtain on various datasets to get a feeling of how it reads. Thx

jasmainak commented:
> There are 6 .fif files tho, so aren't there 6 datasets?

I think they are 6 runs not dataset. To me, a dataset is everything inside bids_root

adam2392 commented:
To fill in:

1. https://openneuro.org/datasets/ds001779/versions/1.0.2


2. https://openneuro.org/datasets/ds002778/versions/1.0.1



3. https://openneuro.org/datasets/ds002904/versions/1.0.0

adam2392 commented:
A few thoughts:

1. participants.tsv summary is usually desired (from all pubs I've seen and written), but it also is difficult to achieve consistency since it is a RECOMMENDED file + format, so optionally, we should be able to summarize report w/o  it.
2. the [BIDS spec version](https://bids-specification.readthedocs.io/en/stable/) is currently 1.4.0, so that can be updated in the code
3. `MEG_TEMPLATE` and `EEG_TEMPLATE` can be pretty easily added at the end of `create_methods_paragraph()`. If someone wants to add that in, the naive thing to do would be just grab different channels from the sidecar.json file I suppose?

jasmainak commented:
@adam2392 can you put [this comment](https://github.com/mne-tools/mne-bids/pull/457#issuecomment-648424227) in the PR description?

adam2392 commented:
> @adam2392 can you put [this comment](https://github.com/mne-tools/mne-bids/pull/457#issuecomment-648424227) in the PR description?

@jasmainak Besides the current breaking of the bids-validator, I was wondering if you could  lmk how the direction currently feels. 

Some notes:
- I can fix rounding very easily to two decimal places.
- We could add a dynamic check on `participants.tsv` file since participants is always a desired summary. But currently, we have no way of determining if the file complies with how we assume the formatting to be (e.g. `M` vs `male` vs `man` etc.). Idk how desirable this is, versus just "switching off the summary of participants because we lack control".
- I added the total # of scans (e.g. total # runs summed across entire bids_root), but didn't update the summaries yet.

jasmainak commented:
> "switching off the summary of participants because we lack control".

not sure I understand this point. But I would say, don't add too much code complexity and branching. It will make life harder for future developers. 

can you add a couple of more examples in the description? At least 6 or 7 in total? Just to get a sense of how the methods paragraph might be useful/handy for researchers. I'll ask a couple of my colleagues to provide feedback what else might be useful to include.

jasmainak commented on mne_bids/report.py:
```diff
@@ -0,0 +1,444 @@
+                            _find_matching_sidecar, _parse_bids_filename,
+                            BIDSPath)
+
+BIDS_DATASET_TEMPLATE = 'The {name} dataset was created ' \
```
use triple quotes to avoid awkward backslash.

jasmainak commented on mne_bids/tests/test_report.py:
```diff
@@ -0,0 +1,86 @@
+
+    agnostic_descrip = 'Data was acquired using a MEG system ' \
+                       '(Elekta manufacturer with line noise at 60 Hz) ' \
+                       'using filters (n/a). Each dataset is ' \
```
I don't know how easy it would be but you could use logic inside these statements to exclude parts of it if it were `n/a`. You can use tempita in mne externals. Look at the mne/report.py file in mne-python.

jasmainak commented:
> Dataset   was created with BIDS version 1.2.2
using MNE-BIDS

We don't really know if it was MNE-BIDS? If so, we should leave it out

adam2392 commented:
> > Dataset   was created with BIDS version 1.2.2
> > using MNE-BIDS
> 
> We don't really know if it was MNE-BIDS? If so, we should leave it out

Is this still true  in the context of #460 ?

sappelhoff commented on mne_bids/config.py:
```diff
@@ -5,7 +5,8 @@
 BIDS_VERSION = "1.4.0"
 
-DOI = 'https://doi.org/10.21105/joss.01896'
+DOI = """https://doi.org/10.21105/joss.01896,
+https://doi.org/10.5281/zenodo.3739558"""
```
I would not add the zenodo DOI here. It's not indexed on Google Scholar and thus doesn't really help anybody.

From the JOSS DOI, one can easily get to Zenodo by clicking on "Software Archive".

Several DOIs for one thing are always problematic, because people get confused what to cite, and then often cite the wrong thing.

jasmainak commented on mne_bids/report.py:
```diff
@@ -0,0 +1,461 @@
+
+
+if __name__ == '__main__':
+    import textwrap
```
I think this should be used in the function above?

jasmainak commented on mne_bids/report.py:
```diff
@@ -0,0 +1,461 @@
+    bids_root = fetch_brainvision_testing_data()
+    bids_root = "/Users/adam2392/Downloads/ds002904-1.0.0"
+    methods_paragraph = create_methods_paragraph(bids_root)
+    print('\n'.join(textwrap.wrap(methods_paragraph, width=50)))
```
80 characters maybe? It's just a nice standard, we could stick to the same :)

jasmainak commented on mne_bids/report.py:
```diff
@@ -0,0 +1,461 @@
+
+    Returns
+    -------
+    template_dict : dict
```
description of what this contains is missing

jasmainak commented on mne_bids/report.py:
```diff
@@ -0,0 +1,461 @@
+    ----------
+    bids_root : str | pathlib.Path
+        The path of the root of the BIDS compatible folder.
+    scans_fpaths : list
```
description missing

jasmainak commented:
@adam2392 take a look at the [datasets tested in MNE-study-template](https://github.com/mne-tools/mne-study-template/blob/master/.circleci/config.yml). Would you mind posting the description for these datasets as well? If we have a substantial number (around 10), we can start to see if this looks good. And then in the study template, you could add a line to add the generated paragraph to the MNE report using [`add_htmls_to_section`](https://mne.tools/stable/generated/mne.Report.html#mne.Report.add_htmls_to_section) so there is an additional layer of testing for MNE-BIDS.

adam2392 commented:
> @adam2392 take a look at the [datasets tested in MNE-study-template](https://github.com/mne-tools/mne-study-template/blob/master/.circleci/config.yml). Would you mind posting the description for these datasets as well? If we have a substantial number (around 10), we can start to see if this looks good. And then in the study template, you could add a line to add the generated paragraph to the MNE report using [`add_htmls_to_section`](https://mne.tools/stable/generated/mne.Report.html#mne.Report.add_htmls_to_section) so there is an additional layer of testing for MNE-BIDS.

Okay, so I took those datasets and running those locally and pasting into the PR description. Also, FYI some of the descriptions aren't up to date w/ some of the minor changes, but I don't want to re-download the datasets locally and redo (They're mainly spelling or grammar issues, or logic on the template string itself, which have been updated). If this is absolutely needed, I can re-download, but it takes a bit of time on my currently home internet :p.

I don't now what you mean by adding it into the study template? Do you mean in the `circleci` of mne-study-template? If so, which file? Logistically, do I just simply add `mne-bids` to the install, generate the report using this PR (once it's merged), and then add the `add_htmls_to_section` function passing in the `mne-bids` generated report?

agramfort commented:
@adam2392 please paste here examples of results so we don't all need to test ourself to give you more feedback. Any dataset is fine but the more the better. we just need to read some examples.

jasmainak commented:
@agramfort there are now [10 examples in the PR description](https://github.com/mne-tools/mne-bids/pull/457#issue-438116145). 

@adam2392 ignore what I said about reports and study templates for now. I'll guide you on this once this PR is done.

jasmainak commented:
@adam2392 can we smooth out the kinks in the following parts:

> (Elekta/Neuromag, /, n, and a manufacturer)

then here:

> filters ({'SpatialCompensation': {'None': []}})

and here:

> The dataset consists of 0 recording sessions .

jasmainak commented on mne_bids/report.py:
```diff
@@ -0,0 +1,461 @@
+{{n_bad}} are removed from analysis). """
+
+IEEG_TEMPLATE = """There are {{n_ecog_chs}} ECoG and
+{{n_seeg_chs}} SEEG channels. """
```
for MEG too, it is useful to say the gradiometers and magnetometers separately.

adam2392 replied:
I got rid of this for now since the modality-agnostic summaries were already quite challenging.

jasmainak commented:
> I think it's due to the fact that the pybids-report still only supports nifti files.

might be helpful to check on an MRI dataset then. Just to get a sense of what are the modality agnostic things they report.

agramfort commented:
```
https://openneuro.org/datasets/ds002778/versions/1.0.1

There are
31 subjects amongst whom there are 0 males and 0
females (0 unknown). There are 0 right hand, 0
left hand and 0 ambidextrous subjects. 

291.998046875 seconds -> 292.0 seconds
1886 channels ???

etc.
```

I think many weirdness in the phrasing. I suggest a
critical look at what is obtained so it works more
convincingly out of the box.

adam2392 commented:
Sorry I think you guys caught me in the midst of me editing the PR description... But minus the few points below, I can re-update the descriptions. Before I do so, I just want to make sure the direction of the PR looks "good" as in the right data components are being shown and in the right way? I can come back to this after we decide it is okay? Mainly because I have to re-copy-paste every time all 10 datasets otherwise.

> @adam2392 can we smooth out the kinks in the following parts:
> 
> > (Elekta/Neuromag, /, n, and a manufacturer)
> 
> then here:
> 
> > filters ({'SpatialCompensation': {'None': []}})
> 
> and here:
> 
> > The dataset consists of 0 recording sessions .

Sure these just need to have some more logic applied.

> https://openneuro.org/datasets/ds002778/versions/1.0.1
> 291.998046875 seconds -> 292.0 seconds

applied rounding logic to 2 decimals.

> 1886 channels ???

made this `1886 recording channels`

> I think many weirdness in the phrasing. I suggest a
> critical look at what is obtained so it works more
> convincingly out of the box.

I was hoping to get a workflow down of what looks "okay" macroscopically, and then fix phrasing. The phrasing is really hard to get right initially, so hoping to do that last based on the "right information" displayed. Lmk if that sounds good.

> > I think it's due to the fact that the pybids-report still only supports nifti files.
> 
> might be helpful to check on an MRI dataset then. Just to get a sense of what are the modality agnostic things they report.

I don't have any on hand, but can take a peek at their report module.

jasmainak commented:
> Mainly because I have to re-copy-paste every time all 10 datasets otherwise.

could you perhaps develop a better workflow for this? E.g., a script that runs the report command for all the datasets and outputs it to a text file that is stored in dropbox. You can share the dropbox link above in the description. It will save you effort.

jasmainak commented:
> 1886 recording channels

but how come 1886 channels? isn't that too large?

adam2392 commented:
> > Mainly because I have to re-copy-paste every time all 10 datasets otherwise.
> 
> could you perhaps develop a better workflow for this? E.g., a script that runs the report command for all the datasets and outputs it to a text file that is stored in dropbox. You can share the dropbox link above in the description. It will save you effort. I think we need to do many iterations on this so copy-pasting will be painful

I like this idea and yeah that sounds better, where I can just re-write on a Dropbox file.... Just spent 30-45 mins redownloading datasets and copy-pasting :(.

adam2392 commented:
> > 1886 recording channels
> 
> but how come 1886 channels? isn't that too large?

This is the total number of recording channels across all scans. I suppose it's not necessary and can be taken out easily, but I have seen it in pubs sometimes.

agramfort commented:
>
> This is the total number of recording channels across all scans. I suppose
> it's not necessary and can be taken out easily, but I have seen it in pubs
> sometimes.
>

it's not very informative. I would report the number of channels per
subject if it varies.

adam2392 commented:
> See output on Dropbox here: https://www.dropbox.com/sh/8yrm9pvq87pq59p/AAChycq2xRR4SegllA0OeGPKa?dl=0

@jasmainak @agramfort 

In the PR description I've tried updating the summary of what is attempted to be summarized. In addition, I've attached a Dropbox link with all the recent files being updated. 

I took out any modality-specific summary for now because it seems to be a bit overstretch. In addition, I took out `emptyroom` for now because not sure how that is even formatted. It gave a bunch of errors when running through sidecar/channels summary because the naming follows a different convention(?).

jasmainak commented:
@adam2392 would you mind concatenating them in one txt file instead of 10? makes it easier to review

adam2392 commented:
> @adam2392 would you mind concatenating them in one txt file instead of 10? makes it easier to review

Just put it into the github PR actually until review is good! Probably easier this way to account for changes over time.

jasmainak commented on report.txt:
```diff
@@ -0,0 +1,120 @@
+316.0 +/- 0.0 total recording channels per scan (316.0 +/- 0.0 are used and 0.0
++/- 0.0 are removed from analysis). The dataset lengths range from 897.08 to
+897.08 seconds, for a total of 897.08 seconds of data recorded over all scans
+(897.08 +/- 0.0).
```
can you also include sampling frequency?

jasmainak commented on report.txt:
```diff
@@ -0,0 +1,120 @@
+subjects are all unknown. The handedness of the subjects are all unknown. The
+ages of the subjects are 25.0 to 25.0 (25.0 +/- 0.0; 1 with unknown age). Data
+was acquired using a MEG system (CTF manufacturer) with line noise at 60 Hz
+using filters ({'SpatialCompensation': {'GradientOrder': '3rd'}}). There are 3
```
the software filters are still broken

jasmainak commented:
I think @hoechenberger might have better attention span than me to review the PR properly :-)

May main suggestion is to make your writing stylistically nicer and resolve corner cases. See some papers to get a sense of what to mention. Here is an example from [an MEG paper](https://www.mitpressjournals.org/doi/pdf/10.1162/jocn_a_01442):

> Electromagnetic brain activity was recorded using the whole-head Elekta Neuromag Vector View306 MEG system (Neuromag Elekta Ltd.) equipped with 102 triple-sensor elements (two orthogonal planar gradi-ometers and one magnetometer per sensor location)

(it would be nice to also mention the number of gradiometers and magnetometers. It's quite common in MEG field)

then

> MEG and EEG (M/EEG) recordings were sampled at 1 kHz and band-pass filtered between 0.03 and 330 Hz. The EOGs (horizontal and vertical eye movements), elec-trocardiograms, and electromyographics were recorded simultaneously with MEG

(you're still missing the sampling frequency)

and 

> Thus, the final sample was composed of 12 participants (seven women, mean age = 24 years). All participants performed six experimental blocks.

(notice how they don't say 24 +/- 5 or something like that. Make it look like it's written professionally for a publication)

Also:

> Signal space separation correction (Taulu & Simola,2006), head movement compensation, and bad channel rejection were done using MaxFilter Software

(you need to extract the keys of the dictionary and make it into sentences like this)

agramfort commented:
@adam2392 I pushed here.

@adam2392 I let you fix SpatialCompensation filters and adding sampling frequency and fix conflicts ok?

we're getting there !

adam2392 commented:
> @adam2392 I pushed here.
> 
> @adam2392 I let you fix SpatialCompensation filters and adding sampling frequency and fix conflicts ok?
> 
> we're getting there !

Awesome thanks! Sorry I'm taking a hiatus because my MBP keyboard and screen broke (dam you butterfly keyboards), and so I've been working on purely logistical paper writing and reading some textbooks in my spare time. Will probably get back on this in 1 week or so.

jasmainak commented:
sounds good, you can blame your keyboard for the fun rebase ;-)

adam2392 commented:
> May main suggestion is to make your writing stylistically nicer and resolve corner cases. See some papers to get a sense of what to mention. Here is an example from [an MEG paper](https://www.mitpressjournals.org/doi/pdf/10.1162/jocn_a_01442):
> (it would be nice to also mention the number of gradiometers and magnetometers. It's quite common in MEG field)

I don't really have the knowledge for where to get these elements, and think the cases could be handled better if someone who uses MEG implements, or discusses on an issue how to. Then I can chime in if needed on how to "append" to the current template. Is that okay? Sorry just don't want to try too much at once.

> (you're still missing the sampling frequency)

Added.

> > Thus, the final sample was composed of 12 participants (seven women, mean age = 24 years). All participants performed six experimental blocks.
> 
> (notice how they don't say 24 +/- 5 or something like that. Make it look like it's written professionally for a publication)

Sorry for the style irregularities. I was focused on including as much information as possible. Regarding `"x +/- y"` vs `(mean x = )`. I feel like just saying the mean age in the summary doesn't say much, whereas the style standard is mean +/- std. I adapted it to reflect this. If it doesn't look good still, lmk.

> Also:
> 
> > Signal space separation correction (Taulu & Simola,2006), head movement compensation, and bad channel rejection were done using MaxFilter Software

Added `SoftwareFilters` to reflect what it should look like in the specification.

adam2392 commented:
In general I think some of the edge cases are way too difficult to account for. For example, if there is just one subject in the dataset, then the grammar looks very poor because it'll say things like:

```
The ages of the subjects ranged from 25 to 25 (mean = 25, std = 0)...
```

Which I think is on the user to just fix those sentences, since accounting for a dataset of size 1 seems overkill. What do you think?

agramfort commented:
```
$ mne_bids report --bids_root ~/mne_data/ds000117
Summarizing participants.tsv ...
Summarizing scans.tsv files...
------------------------------------ REPORT ------------------------------------
The Multisubject, multimodal face processing dataset was created with BIDS
version 1.0.2 by Wakeman, DG, and Henson, RN. This report was generated with
MNE-BIDS (https://doi.org/10.21105/joss.01896). The dataset consists of 16
subjects and 2 recording sessions: meg, and mri. The sex of the subjects
comprised of 9 males and 7 females. The handedness of the subjects were all
unknown. The ages of the subjects ranged from 23.0 to 31.0 (mean = 26.38, std =
2.76; 1 with unknown age). Data was recorded using a MEG system (Elekta/Neuromag
manufacturer) sampled at 1100 Hz with line noise at 50 Hz using
SpatialCompensation. There were 104 scans in total.  For each dataset, there
were on average 404.0 (std = 0.0) recording channels per scan, out of which
404.0 (std = 0.0) were used in analysis (0.0 +/- 0.0 were removed from
analysis).
Traceback (most recent call last):
  File "/Users/alex/work/src/mne-bids/mne_bids/commands/mne_bids_report.py", line 43, in <module>
    run()
  File "/Users/alex/work/src/mne-bids/mne_bids/commands/mne_bids_report.py", line 39, in run
    assert '  ' not in report
AssertionError
```
@adam2392 in my changes I paid attention to avoid issues of extra spaces inserted spuriously. It's not working any more.

agramfort commented:
for the other dataset:

(base) alex@:mne-bids(automethods)$ mne_bids report --bids_root ~/mne_data/ds000248
Summarizing scans.tsv files...
------------------------------------ REPORT ------------------------------------
The ds000248 dataset was created with BIDS version 1.2 by Alexandre Gramfort,
and Matti S Hämäläinen. This report was generated with MNE-BIDS
(https://doi.org/10.21105/joss.01896). The dataset consists of 1 subjects and
Data was recorded using a MEG system (Elekta manufacturer) sampled at 600.61 Hz
with line noise at 60 Hz. There was 1 scan in total. For each dataset, there
were on average 376.0 (std = 0.0) recording channels per scan, out of which
374.0 (std = 0.0) were used in analysis (2.0 +/- 0.0 were removed from
analysis).

@adam2392 have a look at the details and especially transitions between subsentences.
Data -> data.

agramfort commented:
same pb here:

Summarizing participants.tsv ...
Summarizing scans.tsv files...
------------------------------------ REPORT ------------------------------------
The MEG-BIDS Brainstorm data sample dataset was created with BIDS version 1.0.2
by Elizabeth Bock, Peter Donhauser, Francois Tadel, Guiomar Niso, and Sylvain
Baillet. This report was generated with MNE-BIDS
(https://doi.org/10.21105/joss.01896). The dataset consists of 1 subjects and
The sex of the subjects were all unknown. The handedness of the subjects were
all unknown. The ages of the subjects ranged from 25.0 to 25.0 (mean = 25.0, std
= 0.0; 1 with unknown age). Data was recorded using a MEG system (CTF
manufacturer) sampled at 2400 Hz with line noise at 60 Hz using
SpatialCompensation with parameters 3rd GradientOrder. There were 3 scans in
total. Recording durations ranged from 360 to 360 seconds (mean = 360.0, std =
0.0), for a total of 720 seconds of data recorded over all scans. For each
dataset, there were on average 340.0 (std = 0.0) recording channels per scan,
out of which 340.0 (std = 0.0) were used in analysis (0.0 +/- 0.0 were removed
from analysis).

the sentence "The dataset consists of 1 subjects and" is problematic

adam2392 commented:
> @adam2392 in my changes I paid attention to avoid issues of extra spaces inserted spuriously. It's not working any more.

Sorry not sure why that broke, but I just fixed it! 

> the sentence "The dataset consists of 1 subjects and" is problematic

Also fixed this edge case of no sessions in the bids root.

Lmk if there are any other issues, or if I should remove the `report.txt` file and then push up to make the CI green.

agramfort commented on mne_bids/report.txt:
```diff
@@ -0,0 +1,133 @@
+
+
+somato
```
cleanup this file now that we're good

agramfort commented on doc/whats_new.rst:
```diff
@@ -61,6 +61,7 @@ API
 - The :func:`mne_bids.make_bids_basename` function has been updated to create a :code:`mne_bids.utils.BIDSPath` object, which operates like a path object and allows dynamic updating of BIDs entities, by `Adam Li`_ (`#446 <https://github.com/mne-tools/mne-bids/pull/446>`_)
 - The ``datasets.py`` module was removed from ``MNE-BIDS`` and its utility was replaced by ``mne.datasets``, by `Stefan Appelhoff`_ (`#471 <https://github.com/mne-tools/mne-bids/pull/471>`_)
 - :func:`mne_bids.make_dataset_description` now takes the argument `overwrite` which will reset all fields if `True`. If `False`, user-provided fields will no longer be overwritten by :func:`mne_bids.write_raw_bids` when its `overwrite` argument is `True` unless new values are supplied, by `Alex Rockhill`_ (`#478 <https://github.com/mne-tools/mne-bids/pull/478>`_)
+- :func:`mne_bids.make_report` is now available from the `mne_bids` namespace that creates a string output of a summary of the BIDS dataset, by `Adam Li`_ (`#457 <https://github.com/mne-tools/mne-bids/pull/457>`_)
```
document also the new command line tool.

adam2392 commented:
> almost good to go from my end. Can you make the PR title to MRG when all green? thx @adam2392

Not sure why the docs are failing... Do you know if it's something I can fix?

sappelhoff commented on mne_bids/report.py:
```diff
@@ -0,0 +1,552 @@
+      - dataset_description.json file
+      - (optional) participants.tsv file
+      - (optional) modality-agnostic files for (M/I)EEG data,
+        which reads files from the *_scans.tsv file.
```
I think this is the culprit for the red CIs. My suggestion should fix it. @adam2392 

```suggestion
        which reads files from the ``*_scans.tsv`` file.
```

adam2392 commented on mne_bids/tests/test_report.py:
```diff
@@ -0,0 +1,60 @@
+    expected_report = \
+    """This dataset was created with BIDS version 1.4.0 by Please cite MNE-BIDS in your
+publication before removing this (citations in README). This report was
+generated with MNE-BIDS (https://doi.org/10.21105/joss.01896). The dataset
+consists of 1 subjects and 1 recording sessions: 01. The sex of the subjects
+were all unknown. The handedness of the subjects were all unknown. The ages of
+the subjects all unknown. Data was recorded using a MEG system (Elekta
```
It's probably possible, and yeah I can get behind the terminology suggestion.

Before I implement, just want to make sure everyone on board? @jasmainak @agramfort

adam2392 replied:
Jk not that hard, so pushed it up :P.

agramfort commented:
I like what Richard suggests

adam2392 commented:
One issue I'm having is... is there a way one can "add" `Template` objects together? It would make handling participants string easier.

jasmainak commented:
please go ahead and incorporate @hoechenberger and @sappelhoff 's suggestions :-) I am lagging behind ...

adam2392 commented:
Okay I added in the feedback and I think things are all good to go.

agramfort commented:
here is what I get

```
(base) alex@:mne-bids(automethods)$ mne_bids report --bids_root ~/mne_data/ds000117/
Summarizing participants.tsv /Users/alex/mne_data/ds000117/participants.tsv...
Summarizing scans.tsv files [PosixPath('/Users/alex/mne_data/ds000117/sub-13/ses-meg/sub-13_ses-meg_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-14/ses-meg/sub-14_ses-meg_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-emptyroom/ses-20090506/sub-emptyroom_ses-20090506_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-emptyroom/ses-20090601/sub-emptyroom_ses-20090601_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-emptyroom/ses-20090515/sub-emptyroom_ses-20090515_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-emptyroom/ses-20091208/sub-emptyroom_ses-20091208_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-emptyroom/ses-20090409/sub-emptyroom_ses-20090409_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-emptyroom/ses-20091126/sub-emptyroom_ses-20091126_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-emptyroom/ses-20090518/sub-emptyroom_ses-20090518_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-emptyroom/ses-20090511/sub-emptyroom_ses-20090511_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-15/ses-meg/sub-15_ses-meg_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-12/ses-meg/sub-12_ses-meg_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-08/ses-meg/sub-08_ses-meg_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-01/ses-meg/sub-01_ses-meg_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-06/ses-meg/sub-06_ses-meg_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-07/ses-meg/sub-07_ses-meg_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-09/ses-meg/sub-09_ses-meg_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-10/ses-meg/sub-10_ses-meg_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-11/ses-meg/sub-11_ses-meg_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-16/ses-meg/sub-16_ses-meg_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-05/ses-meg/sub-05_ses-meg_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-02/ses-meg/sub-02_ses-meg_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-03/ses-meg/sub-03_ses-meg_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000117/sub-04/ses-meg/sub-04_ses-meg_scans.tsv')]...
The participant template found: comprised of 9 men and 7 women;
handedness were all unknown; ages ranged from 23.0 to 31.0 (mean = 26.38, std = 2.76; 1 with unknown age)
------------------------------------ REPORT ------------------------------------
The Multisubject, multimodal face processing dataset was created with BIDS
version 1.0.2 by Wakeman, DG, and Henson, RN. This report was generated with
MNE-BIDS (https://doi.org/10.21105/joss.01896). The dataset consists of 16
participants (comprised of 9 men and 7 women; handedness were all unknown; ages
ranged from 23.0 to 31.0 (mean = 26.38, std = 2.76; 1 with unknown age))and 2
recording sessions: meg, and mri. Data was recorded using a MEG system
(Elekta/Neuromag manufacturer) sampled at 1100 Hz with line noise at 50 Hz using
SpatialCompensation. There were 104 scans in total. For each dataset, there were
on average 404.0 (std = 0.0) recording channels per scan, out of which 404.0
(std = 0.0) were used in analysis (0.0 +/- 0.0 were removed from analysis).
(base) alex@:mne-bids(automethods)$ mne_bids report --bids_root ~/mne_data/ds000248
Summarizing scans.tsv files [PosixPath('/Users/alex/mne_data/ds000248/sub-01/sub-01_scans.tsv')]...
------------------------------------ REPORT ------------------------------------
The ds000248 dataset was created with BIDS version 1.2 by Alexandre Gramfort,
and Matti S Hämäläinen. This report was generated with MNE-BIDS
(https://doi.org/10.21105/joss.01896). The dataset consists of 1 participants
(). Data was recorded using a MEG system (Elekta manufacturer) sampled at 600.61
Hz with line noise at 60 Hz. There was 1 scan in total. For each dataset, there
were on average 376.0 (std = 0.0) recording channels per scan, out of which
374.0 (std = 0.0) were used in analysis (2.0 +/- 0.0 were removed from
analysis).
(base) alex@:mne-bids(automethods)$ mne_bids report --bids_root ~/mne_data/ds000246
Summarizing participants.tsv /Users/alex/mne_data/ds000246/participants.tsv...
Summarizing scans.tsv files [PosixPath('/Users/alex/mne_data/ds000246/sub-emptyroom/sub-emptyroom_scans.tsv'), PosixPath('/Users/alex/mne_data/ds000246/sub-0001/sub-0001_scans.tsv')]...
The participant template found: sex were all unknown;
handedness were all unknown; ages ranged from 25.0 to 25.0 (mean = 25.0, std = 0.0; 1 with unknown age)
------------------------------------ REPORT ------------------------------------
The MEG-BIDS Brainstorm data sample dataset was created with BIDS version 1.0.2
by Elizabeth Bock, Peter Donhauser, Francois Tadel, Guiomar Niso, and Sylvain
Baillet. This report was generated with MNE-BIDS
(https://doi.org/10.21105/joss.01896). The dataset consists of 1 participants
(sex were all unknown; handedness were all unknown; ages ranged from 25.0 to
25.0 (mean = 25.0, std = 0.0; 1 with unknown age)). Data was recorded using a
MEG system (CTF manufacturer) sampled at 2400 Hz with line noise at 60 Hz using
SpatialCompensation with parameters 3rd GradientOrder. There were 3 scans in
total. Recording durations ranged from 360 to 360 seconds (mean = 360.0, std =
0.0), for a total of 720 seconds of data recorded over all scans. For each
dataset, there were on average 340.0 (std = 0.0) recording channels per scan,
out of which 340.0 (std = 0.0) were used in analysis (0.0 +/- 0.0 were removed
from analysis).
```

jasmainak commented:
Looks pretty fair!

jasmainak commented:
We can keep tweaking more in future PRs but this is good for a start.

agramfort commented:
ok for you @hoechenberger ?

adam2392 commented:
Any remaining changes needed here? Don't want this to go stale and we forget what happened :p.

agramfort commented:
maybe we merge this and improve later?

sappelhoff commented:
> maybe we merge this and improve later?

+1, I hate stale PRs and love iterations :-)

jasmainak commented:
Merged, thanks a ton @adam2392 !!! This is great :-)

hoechenberger commented:
Argh I was just working on a review… but yet let's iterate, then! :)

jasmainak commented:
oops, sorry :)

On Thu, Jul 30, 2020 at 5:33 PM Richard Höchenberger <
notifications@github.com> wrote:

> Argh I was just working on a review… but yet let's iterate, then! :)
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-bids/pull/457#issuecomment-666717601>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADY6FIVRSFJCZI2J3PRVAQ3R6HRKJANCNFSM4OE6BDQQ>
> .
>
